### PR TITLE
WNet + models code refactor

### PR DESCRIPTION
**Pull request by C-Achard**
PR adding WNet self-supervised model, and overhauled code for models usage

Note : merge to main after #34 has been merged (branch dependency)

- [x] Run new pre-commit hooks from this branch on main before merging to avoid conflicts
- [x] Update docs for custom model template
- [x] Document WNet usage
- [x] Add notebook for training 
-> WILL BE DONE IN OTHER PR : Improve ONNX support/UI/docs ~~and add pre-trained ONNX model to HF~~

### CODE REVIEW

**Review by nastya236 — COMMENTED** [review]
In my opinion, the code is well written. + documentation to all classes!! 
Well done! I commented the places in the code which could be (from my personal perspective) done a little bit more flexible.

**Comment by MMathisLab** [comment]
conflicts would be need fixed, and docs! then looks good :D

**Comment by C-Achard** [comment]
@MMathisLab Working on rebase now ! Thanks for the review. And yes, docs will be added, just thought I'd open the code for review

**Review by MMathisLab — APPROVED** [review]

File: docs/res/guides/custom_model_template.rst [hunk 1]
```diff
@@ -3,35 +3,33 @@
 Advanced : Declaring a custom model
 =============================================
 
-To add a custom model, you will need a **.py** file with the following structure to be placed in the *napari_cellseg3d/models* folder:
+.. warning::
+    **WIP** : Adding new models is still a work in progress and will likely not work simply by adding the model in the plugin.
```
Comment (CYHSM) [inline docs/res/guides/custom_model_template.rst:7]: The last part of this sentence is a bit confusing. I think you can just write "Adding new models is still a work in progress and will likely lead to errors".

File: docs/res/guides/custom_model_template.rst [hunk 2]
```diff
@@ -3,35 +3,33 @@
...
+            pass
 
 
-    def get_validation(model, val_inputs):
-        val_outputs = model(val_inputs) # should return the proper type for validation
-        # with sliding_window_inference from MONAI
-        return val_outputs
-
+.. note::
+    **WIP** : Currently you must modify :ref:`model_framework.py` as well : import your model class and add it to the ``model_dict`` attribute
```
Comment (CYHSM) [inline docs/res/guides/custom_model_template.rst:33]: Whats the reason for also changing model_framework?

File: napari_cellseg3d/_tests/test_plugin_inference.py [hunk 3]
```diff
@@ -28,14 +34,34 @@ def test_inference(make_napari_viewer, qtbot):
...
+    worker.config.post_process_config.instance.method = (
+        INSTANCE_SEGMENTATION_METHOD_LIST["Watershed"]()
+    )
+
+    assert worker.config.layer is not None
+    worker.log_parameters()
+
+    res = next(worker.inference())
+    assert isinstance(res, InferenceResult)
+    assert res.result.shape == (8, 8, 8)
```
Comment (CYHSM) [inline napari_cellseg3d/_tests/test_plugin_inference.py:65]: This is the result for the particular window size, i.e. 64? Could add some test for edge cases (not sure if needed though)

File: napari_cellseg3d/code_models/models/wnet/model.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    """Input block of the U-Net architecture."""
+
+    def __init__(self, in_channels, out_channels, dropout=0.65):
+        super(InBlock, self).__init__()
+        # self.device = device
+        self.module = nn.Sequential(
+            nn.Conv3d(in_channels, out_channels, 3, padding=1),
+            nn.ReLU(),
+            nn.Dropout(p=dropout),
+            nn.BatchNorm3d(out_channels),
```
Comment (CYHSM) [inline napari_cellseg3d/code_models/models/wnet/model.py:178]: I wonder why they use BatchNorm instead of Layer or GroupNorm. Do you know if they made any ablation studies on that?

File: napari_cellseg3d/code_models/models/wnet/soft_Ncuts.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+            denominator = torch.sum(
+                class_probs * F.conv3d(weights, kernel, padding=self.radius),
+                dim=(1, 2, 3, 4),
+            )
+            loss += nn.L1Loss()(
+                numerator / torch.add(denominator, 1e-6),
+                torch.zeros_like(numerator),
+            )
+
+        return K - loss
```
Comment (CYHSM) [inline napari_cellseg3d/code_models/models/wnet/soft_Ncuts.py:111]: wait, the overall loss is the number of classes - loss?

**Review by CYHSM — APPROVED** [review]
Code looks good and very clean to me, just added some comments (can also discuss later)

File: docs/res/guides/custom_model_template.rst (hunk 2 above)
Comment (C-Achard) [reply docs/res/guides/custom_model_template.rst:33]: Basically, it's needed so it shows up in the menus. IIRC adding it to config (which would be more centralized and arguably better UX) leads to circular import issues so I moved it there

File: docs/res/guides/custom_model_template.rst (hunk 1 above)
Comment (C-Achard) [reply docs/res/guides/custom_model_template.rst:7]: Done !

File: napari_cellseg3d/_tests/test_plugin_inference.py (hunk 3 above)
Comment (C-Achard) [reply napari_cellseg3d/_tests/test_plugin_inference.py:65]: Sorry, I'm not sure what you mean ? Input is 6^3, padded to 8^3, hence the check.
Edge cases should be handled, since :
- If window is disabled, data is padded up to nearest power of 2
- If window is enabled, MONAI should take care of making the dimensions correct

File: napari_cellseg3d/code_models/models/wnet/soft_Ncuts.py (hunk 5 above)
Comment (C-Achard) [reply napari_cellseg3d/code_models/models/wnet/soft_Ncuts.py:111]: Yes, that's why it goes up by 1 when you add classes. In our case it means that it's always bounded between 0 and 1

Comment (C-Achard) [reply napari_cellseg3d/code_models/models/wnet/soft_Ncuts.py:111]: It's due to needing the Soft version for backward prop

File: napari_cellseg3d/_tests/test_plugin_inference.py (hunk 3 above)
Comment (C-Achard) [reply napari_cellseg3d/_tests/test_plugin_inference.py:66]: (Added an extra in the test to cover this)

File: docs/res/guides/custom_model_template.rst (hunk 2 above)
Comment (C-Achard) [reply docs/res/guides/custom_model_template.rst:33]: @CYHSM We could auto-"scan" for classes in the models folder though, but it's not super hard to edit rn. And if you're adding models you'd know how to import your class and add it to the dict I guess
